### DISCUSSION TITLE
[Chapter 08] 템플릿 메서드 패턴과 전략 패턴

### DISCUSSION THREAD

**Discussion by corock** [cowaine/is-my-code-that-strange#59] [Chapter 07-08 (컬렉션, 강한 결합)] [answered]
지난번 6장에서는 종류별로 다르게 처리해야 하는 기능을 인터페이스의 메서드로 정의하는 전략 패턴을 설명하고 예시와 함께 리팩터링했습니다.

> 6.2.7 인터페이스를 `switch` 조건문 중복에 응용하기(전략 패턴) (_p. 105_)

한편, 8장에서는 상속을 사용하는 나쁜 일반화를 소개하면서 아래와 같이 템플릿 메서드 패턴을 언급하고 있습니다.

> 상속받는 쪽에서 차이가 있는 로직만 구현하는 템플릿 메서드(_template method_)라는 디자인 패턴도 있습니다. (_p. 167_)

많은 도서와 강의, 심지어 학부 시절 수강했던 디자인 패턴 강의 교수님마저도 위 두 가지 패턴을 비교할 정도로 둘은 비슷한 구석이 있어 보입니다.
소위 말하는 (디자인) 패턴 무새는 되지 않으려고 하지만, 책에서 언급하기도 했고 학습 차원에서 이번 기회에 정리할 수 있는 시간이 되었으면 합니다.

궁금한 점은 다음과 같습니다.

1. 템플릿 메서드 패턴과 전략(_strategy_) 패턴은 왜 자주 비교되는가? 어떤 비슷한 점이 있어서?
2. 템플릿 메서드는 무엇인가? (전략 패턴은 책에서 소개하므로 SKIP)
3. 잘 알려진 프레임워크(_e.g._ 스프링)나 오픈 소스에서 템플릿 메서드 적용 사례 또는 본인이 직접 구현했던 경험

**Comment by firewoody237** [comment, marked as answer]
저는 아직 패턴자체를 깊게 공부하지는 않고 개념만 알고있는데요,
(때문에 틀릴수도 있습니다 ㅎㅎ)

### 템플릿 메서드 패턴과 전략(strategy) 패턴은 왜 자주 비교되는가? 어떤 비슷한 점이 있어서?
아래와 같은 차이가 있다고 생각됩니다.
- 템플릿 메서드 패턴
    - 템플릿 자체가 default method와 같은 어떤 형태로 지정되어있다. (변경될 메서드의 앞과 뒤 로직을 전부 관리한다.)
    - 구현객체의 메서드에서, 템플릿은 유지한 채 내부의 일부 실행메서드만 다르게 적용한다.
- 전략 패턴
    - 변경될 메서드에 앞과 뒤 로직과 상관 없이, 특정 메서드의 전략만 변경한다.
- 따라서 템플릿 메서드 패턴에 비해 비교적 자유롭다고 생각합니다.

### 잘 알려진 프레임워크(e.g. 스프링)나 오픈 소스에서 템플릿 메서드 적용 사례 또는 본인이 직접 구현했던 경험
- Jdbc Template
- AsyncTaskExecutor(구현체의 `execute()`)

**Comment by Irisation23** [comment]
### 금일 CheckPoint

1. 템플릿 메서드 패턴과 전략(strategy) 패턴은 정보 은닉화를 기반으로 만들어진 디자인 패턴이다.
2. https://engineering.linecorp.com/ko/blog/templete-method-pattern

**Comment by snoopingh** [comment]
전혀 다른 목적의 패턴이라고 생각했는데, 두 패턴 모두 "변하는 부분과 변하지 않는 부분을 정의하고 코드의 재사용성과 유지 보수성을 높인다" 는 점에서 비슷한 부분이 있기 때문에 비교된다는 것 같네요.

템플릿 메소드 패턴
변하지 않는 부분 : 전체적인 흐름이나 뼈대는 추상 클래스에서 정의된 템플릿 메서드에 의해 고정. (단계들의 순서와 구조는 변경 x)
변하는 부분 : 특정 단계는 추상 메소드로 정의되어, 서브 클래스에서 이를 구현

전략 패턴
변하지 않는 부분 : 알고리즘을 수행하는 기본적인 구조를 정의한 컨텍스트 클래스
변하는 부분 : 전략 인터페이스와 이를 구현한 구체적인 전략 클래스들은 런타임시에 교체될 수 있음
